OLX export: Reverted url_name from a constant to course run

amir-qayyum-khan opened the pull request:
#### What are the relevant tickets?
fixes https://github.com/mitodl/edx-platform/issues/31

#### Background
[Nimisha Asthagiri views](https://openedx.atlassian.net/browse/CRI-102?focusedCommentId=329723&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-329723)

I agree that OLX needs a lot of TLC. I just don't see the "side-car" concept as something that moves us forward in a direction that makes sense for the future - given what we need for OLX and decoupling content and enabling reuse.

What I'm proposing is this:

Update the course export code for Split courses so that the value of course-run is placed in the same place as it is for non-Split courses: in the url_name field in the course.xml file.
On OLX import, make sure the "right" thing happens for Split courses. It may already be given the code that exists in import_xblock in Split. That import code sets the "block_id" to "course" for the course_module.
This would be a low-cost solution, without inventing new OLX concepts that we may not need. As another advantage, the resulting OLX would be consistent with non-Split courses.

#### What's this PR do?
- It add course run as child directory inside policies folder.
- it updates `policy.json` file and add `course/{course_run}` as root element
 
#### How should this be manually tested?

1. Create a course in Studio with a unique course run, e.g. the suggested `2104_T1`
2. Export the course and examine the export:
    1. In the root course.xml the value of `course_xml` should be the run you set in studio, not `course`.
    2. In the policies directory, there should be a sub-directory named after the run. Not `course`. 
    3. In policy.json file, the root key should be named `course/[run]` not `course\course`
3. import the course into lms using the management command, for example
  `python /edx/app/edxapp/edx-platform/manage.py lms import /edx/src/course/ /edx/src/course/`
    1. Confirm that the count of courses are the same -- importing the course didn't create a new one with a course run of `course`

edx-webhook commented:
Thanks for the pull request, @amir-qayyum-khan! I've created [OSPR-2570](https://openedx.atlassian.net/browse/OSPR-2570) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

pdpinch commented:
@amir-qayyum-khan this is working as I expected and I think the code approach is correct. You have some tests that will need to be updated though.

amir-qayyum-khan commented:
Done @pdpinch, please take a look.

pdpinch commented on common/lib/xmodule/xmodule/x_module.py:
```diff
@@ -1116,7 +1116,12 @@ def add_xml_to_node(self, node):
         node.tail = exported_node.tail
+
         for key, value in exported_node.items():
+            print key, value
```
debugging?

amir-qayyum-khan replied:
oops, committed by mistake

pdpinch commented on common/lib/xmodule/xmodule/x_module.py:
```diff
@@ -1116,7 +1116,12 @@ def add_xml_to_node(self, node):
+            print key, value
+            if key == 'url_name' and value == 'course' and key in node.attrib:
+                # url_name for this node is already set to course run
+                continue
```
I don't understand this comment. Should the value for `url_name` be the course run, instead of `course`?

amir-qayyum-khan replied:
@pdpinch i am setting value to course run in `ExportManager`. 
This code states if the value is set in `ExportManager` then skip it setting to `course`. Otherwise set it what ever you like `course` or any key.

amir-qayyum-khan replied:
update my comment

pdpinch replied:
thanks, that's much clearer.

pdpinch commented:
@amir-qayyum-khan one small comment and a question

amir-qayyum-khan commented:
Addressed @pdpinch , please review it again

mduboseedx commented:
@amir-qayyum-khan Thank you for the contribution and I apologize for the delay as I was out of the office last week. Feel free to let me know when this is ready for a review.

pdpinch commented:
@mduboseedx this is ready for review from my perspective. @ormsbee or @nasthagiri may have advice on who should review.

ormsbee commented:
@mduboseedx, @pdpinch: I can take this one.

amir-qayyum-khan commented:
> What happens when you try to take a course that was exported in this manner and import it into an older version of edx-platform?

No change in import feature, Import feature is generic. it take value of url_name and process it. So when i changed `url_name` to course run i had to rename some files and folder because import expect files in that way.

- rename course.xml to {course run}.xml in course folder because its name must be same as url_name value
- changed directory in policies to course run because import expect its name to be equal to value of url_name
- changed first key in policy.json because again it should be same as course run


> Have we changed a file name in the export? If so, do we need to do anything for backwards compatibility (e.g. export the old way in addition to the new)?

I dont think so. Import code is generic and it works on url_name value. So what ever we export, should follow this principle. If yes imported will import it.

> Does this have any effect on Content Libraries?

i dont think so.

> Whatever the answers are to the above questions, can it also be reflected in the commit message of this PR?
sure, let me try

cc @pdpinch

pdpinch commented:
I agree with Amir's answers -- but I'd like to get a chance to review the commit message. 

@ormsbee do you want us to say something more definitive about backward compatibility, like "courses exported with this code have been tested to import correctly into [hawthorn|ginko|ficus]". How far back would you like us to test?

pdpinch commented:
@ormsbee this is what I'm thinking for the commit message. Let me know if this is off-base. If I don't hear from you, I'll assume it's OK. 

```
OLX export: Reverted url_name from a constant to course run

- New exported courses include course run information in:
  - `url_name` of root course node
  - file name of root node in course folder
  - root key name in policy.json
  - directory name inside policies folder

- when imported via management command, the OLX will overwrite an available existing course with the same course key (i.e. same org, course number and course run)
  - if there is no matching course, one will be created

- when imported via the studio web ui (or import API), the OLX will replace the current course (no change in behavior)

- courses exported with this commit have been tested to import via management command and studio web UI in hawthorn and ginkgo releases. They should also work in prior releases, but have not been tested.
```

ormsbee commented:
@pdpinch: LGTM

amir-qayyum-khan commented:
jenkins run bokchoy

pdpinch commented:
Thanks. Just to double-check, in scenario 2, the run information was maintained, right? 

In scenario 4, I would expect it to be lost (all the way back to ficus, I think).

amir-qayyum-khan commented:
@pdpinch this code is all about `url_name`. if value of this is to course run then import module expects same files and folder to be name as `url_name==course run`. if yes the import on ficus with use url_name as course run. It will not change it in import.

pdpinch commented:
Ok. That's what I was expecting. I'm going to merge this. fyi @ormsbee

edx-webhook commented:
@amir-qayyum-khan 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/18840&entry.1484655318=https://github.com/amir-qayyum-khan&entry.752974735=2018-08-27+11:59&entry.1917517419=2018-09-14+14:00&entry.2133058324=Yes) so we can improve your experience in the future.
